src/program/latex/mod.rs

```diff
@@ -0,0 +1,1600 @@
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+
+    /// Returns a reformatted gate name based on the modifiers used in a single
+    /// instruction line of a quil program or the original name. Gates with
+    /// CONTROLLED modifiers are reformatted such that each CONTROLLED modifier
+    /// prepends a `C` to the beginning of the gate name. For other modifiers
+    /// such as DAGGER, no special reformatting is required.
+    ///
+    /// For example, for an instruction line `CONTROLLED CONTROLLED Y 0 1 2` the
+    /// gate name is reformatted to `CCY` where each C is mapped to an
+    /// associated qubit with the last qubit to the original gate name. For an
+    /// instruction line `DAGGER DAGGER Y 0`, the gate name remains `Y`,
+    /// instead each of the modifiers are added to the wire at the current
+    /// column where it is to be applied using the Command::Super variant.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the current column of the Circuit
+    /// `gate` - exposes the modifiers associated with the instruction gate
+    /// `wire` - exposes the wire to be pushed to in Circuit
+    fn set_modifiers(&mut self, gate: &Gate, wire: &mut Wire) -> String {
+        let mut gate_name = gate.name.clone();
+
+        // set modifers
+        if !gate.modifiers.is_empty() {
+            for modifer in &gate.modifiers {
+                match modifer {
+                    instruction::GateModifier::Dagger => {
+                        if let Some(modifiers) = wire.modifiers.get_mut(&self.column) {
+                            modifiers.push("dagger".to_string())
+                        } else {
+                            wire.modifiers
+                                .insert(self.column, vec!["dagger".to_string()]);
+                        }
+                    }
+                    instruction::GateModifier::Controlled => {
+                        // prepend a C to the gate
+                        gate_name.insert_str(0, "C");
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        gate_name
+    }
+
+    /// For every instruction containing control and target qubits this method
+    /// identifies which qubit is the target and which qubit is controlling it.
+    /// The logic of this function is visualized using a physical vector with
+    /// the tail at the control qubit and the head pointing to the target
+    /// qubit. The distance between the qubits represents the number of wires
+    /// between them, i.e the space that the vector needs to traverse. If the
+    /// control qubit comes before the target qubit the direction is positive,
+    /// otherwise, it is negative. See [`Quantikz`] documentation on CNOT for
+    /// some background that helps justify this approach.
+    ///
+    /// This function is expensive with a time complexity of O(n^2). In the
```

Comment:
We may be able to avoid this costly function by determining the size of the circuit up front. If we know how many wires/rows the circuit will have, then the `ctrl` and `targ` values can be calculated as the difference between the qubit indices for the gate.